Add upgrade support to CashTrickler and split off GivesCashOnCapture

reaperrr opened the pull request:
Supersedes #11836.

Closes #11835.

Commit:
Revamp CashTrickler.

Added upgrade support.
Decoupled the capture bonus to a separate trait.

reaperrr committed:
CashTrickler style fixes and Period -> Interval rename

Phrohdoh commented on OpenRA.Mods.Common/Traits/GivesCashOnCapture.cs:
```diff
@@ -0,0 +1,57 @@
+		void MaybeAddCashTick(Actor self, int amount)
+		{
+			if (info.ShowTicks)
+				self.World.AddFrameEndTask(w => w.Add(new FloatingText(self.CenterPosition, self.Owner.Color.RGB, FloatingText.FormatCashTick(amount), 30)));
```
Might as well unhardcode the `30` since you're here.

Phrohdoh commented on OpenRA.Mods.Common/Traits/GivesCashOnCapture.cs:
```diff
@@ -0,0 +1,57 @@
+			if (IsTraitDisabled)
+				return;
+
+			if (info.CaptureAmount > 0)
```
This should probably be a lint as a value `<= 0` on this trait doesn't make any sense.

reaperrr replied:
Actually, I think `> 0` shouldn't ben enforced at all, neither in code nor via lint error, because some modders might want to use negative values for minigames and such. A mere lint warning should be enough.

pchote replied:
Allowing negative values seems reasonable, but you'll need to make sure it behaves correctly when the player has less cash than it wants to take.

reaperrr committed:
Safeguard CashTrickler negative values

By making sure to never remove more cash than the player has.

reaperrr commented:
Updated.

WolfGaming commented on OpenRA.Mods.Common/Traits/GivesCashOnCapture.cs:
```diff
@@ -0,0 +1,78 @@
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Lets the actor grant cash when captured.")]
+	class GivesCashOnCaptureInfo : UpgradableTraitInfo
```
Just a thought why not just CashOnCapture?

reaperrr replied:
`GivesCashOnCapture` is more in line with `GivesExperience` and `GivesBounty`. I know there are contrary examples, but my point is that this is consistent enough to not require another rename, in my opinion.

WolfGaming commented on OpenRA.Mods.Common/Traits/CashTrickler.cs:
```diff
@@ -15,52 +15,73 @@
+
+			// Check whether the amount of cash to be removed would exceed available player cash, in that case only remove all the player cash
+			var newAmount = totalCash + amount < 0 ? -totalCash : amount;
+			resources.TakeCash(-newAmount);
```
Can you take this - away from newAmount and just take in a positive number into RemoveCash, you are doing a lot of gymnastics here to have the right sign in front of the number.

reaperrr commented:
Updated.

WolfGaming commented:
Did the implementation before include the option of disabling the capture cash after the first time the building is captured or?
Don't seem to find this in the code you removed so unless that is a problem this gets a 👍 from me.

GraionDilach commented:
@WolfGaming Introducing that feature was among the intentions of the code, providing an upgrade through GlobalUpgradable for all players except Neutral and Creeps achieves that.

pchote commented on OpenRA.Mods.Common/Traits/CashTrickler.cs:
```diff
@@ -15,52 +15,73 @@
+
+		void AddCash(Actor self, Player owner, int amount)
+		{
+			owner.PlayerActor.Trait<PlayerResources>().GiveCash(amount);
```
Not a big deal, but you could save trait lookups by caching this at create time and updating the reference on capture.

pchote commented on OpenRA.Mods.Common/Traits/CashTrickler.cs:
```diff
@@ -15,52 +15,73 @@
+			var totalCash = resources.Cash + resources.Resources;
+
+			// Check whether the amount of cash to be removed would exceed available player cash, in that case only remove all the player cash
+			var newAmount = totalCash - amount < 0 ? totalCash : amount;
```
This can be condensed down to one line with `var newAmount = Math.Min(resources.Cash + resources.Resources, amount);`

pchote commented on OpenRA.Mods.Common/Traits/GivesCashOnCapture.cs:
```diff
@@ -0,0 +1,78 @@
+		public readonly int DisplayDuration = 30;
+
+		[Desc("Amount of money awarded for capturing the actor.")]
+		public readonly int CaptureAmount = 0;
```
The `Capture` isn't needed anymore, so please rename this to `Cash` (to make it clear that it is not giving resources)

pchote commented on OpenRA.Mods.Common/Traits/GivesCashOnCapture.cs:
```diff
@@ -0,0 +1,78 @@
+		void AddCash(Actor self, Player newOwner, int amount)
+		{
+			newOwner.PlayerActor.Trait<PlayerResources>().GiveCash(amount);
+			if (info.ShowTicks)
```
This can move to the end of `OnCapture` to avoid duplication.

pchote commented on OpenRA.Mods.Common/Traits/GivesCashOnCapture.cs:
```diff
@@ -0,0 +1,78 @@
+				AddCash(self, newOwner, info.CaptureAmount);
+		}
+
+		void AddCashTick(Actor self, int amount)
```
This will only be used once, so can be moved inline.

reaperrr commented:
Rebased and updated.

I condensed the Add/RemoveCash methods into a single ModifyCash method, that reduces duplication and makes the code a bit easier to review.

Phrohdoh commented:
Would anyone (proper mods, thirdparty) expect a cash-tick of `$0`?
If not (and I expect this to be the case) you could bail out of `ModifyCash` early if `amount == 0` and avoid lookups.

pchote commented on OpenRA.Mods.Common/Traits/CashTrickler.cs:
```diff
@@ -9,58 +9,78 @@
 		{
-			if (info.ShowTicks)
-				self.World.AddFrameEndTask(w => w.Add(new FloatingText(self.CenterPosition, self.Owner.Color.RGB, FloatingText.FormatCashTick(amount), 30)));
+			var resources = newOwner.PlayerActor.Trait<PlayerResources>();
```
Can we cache this instead of looking it up each cycle?  You can refer to the `Refinery` trait to see an example of how to set and change it on ownership change.

pchote commented on OpenRA.Mods.Common/Traits/GivesCashOnCapture.cs:
```diff
@@ -0,0 +1,71 @@
+		public readonly int DisplayDuration = 30;
+
+		[Desc("Amount of money awarded for capturing the actor.")]
+		public readonly int Cash = 0;
```
Can we call this `Amount` to match `CashTrickler`?

pchote commented on OpenRA.Mods.Common/Traits/GivesCashOnCapture.cs:
```diff
@@ -0,0 +1,71 @@
+				return;
+
+			self.World.AddFrameEndTask(w => w.Add(
+				new FloatingText(self.CenterPosition, self.Owner.Color.RGB, FloatingText.FormatCashTick(amount), info.DisplayDuration)));
```
This tick value will be wrong if the `drain` is capped above.

reaperrr replied:
Adressed in linr 58.

pchote commented on OpenRA.Mods.Common/Traits/GivesCashOnCapture.cs:
```diff
@@ -0,0 +1,71 @@
+			ModifyCash(self, newOwner, info.Cash);
+		}
+
+		void ModifyCash(Actor self, Player newOwner, int amount)
```
Please fold this into the method above, there's no reason to have it separate (having duplicated code blocks across different traits isn't a great goal).

reaperrr commented:
Updated and rebased.

GraionDilach commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -368,8 +368,9 @@ internal static void UpgradeActorRules(ModData modData, int engineVersion, ref L
 					}
 				}
 
+				// Update rules added during prep-1609 stable period, dates need fixing after release.
```
Probably drop this change now that you splitted it off?

GraionDilach reviewed (commented):
Dunno how much I can :+1: my own code this way but looks good to me.

reaperrr commented:
Updated.

> Dunno how much I can 👍 my own code this way but looks good to me.

The code changed quite a bit since your original PR, so in my opinion it counts.

obrakmann commented on OpenRA.Mods.Common/Traits/GivesCashOnCapture.cs:
```diff
@@ -0,0 +1,70 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2016 The OpenRA Developers (see AUTHORS)
```
2017, please

obrakmann reviewed (approved):
👍 after the header has been adjusted.

reaperrr commented:
Fixed.

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/0a36f868a55161e6b4d87235f2ec63e3ca1b57e0)
